### ISSUE TITLE
west update fails on submodules with relative paths

### ISSUE THREAD

**Issue by Mierunski** [zephyrproject-rtos/west#545]
When west initializes project repository it only adds remote with name specified in `west.yml` while git assumes that default remote is `origin`. Relative submodule paths are resolved to default remote. 
Without `origin` remote, `west update` fails while trying to update repository that contains submodules with relative paths, with following warning
 ```
warning: could not look up configuration 'remote.origin.url'. Assuming this repository is its own authoritative upstream.
```
git logic for resolving relative paths can be found here https://github.com/git/git/blob/v2.20.1/builtin/submodule--helper.c#L135
Manually adding `origin` remote solves the issue.
i'd propose west should also add `origin` remote to projects by default.

**Comment by marc-hb** [comment]
Good catch. As a coincidence I did notice recently (outside west) that relative submodules paths indeed assume there is an `origin`. I mean: issue confirmed.

@mbolivar-nordic can you remind us why `west` uses `upstream` by default  and why this can't be changed? Sorry I forgot :/

**Comment by marc-hb** [comment]
> > repository that contains submodules with relative paths

> I don't know what this means. @Mierunski can you please provide a manifest and steps to reproduce?

It's simply a .gitmodules file without a hostname. Useful for mirroring. Here's one example: https://github.com/thesofproject/sof/commit/250ab3875c23

> The value is determined by zephyr/west.yml in this case. https://github.com/zephyrproject-rtos/zephyr/blob/045c4bbf72d358db320c23be1caa3b5564fb3e2c/west.yml#L22

@Mierunski can you please change your west.yml and replace `upstream` with `origin` and see if this avoids this issue? I'm not trying to dismiss your problem, for now I'm only interested in proving @mbolivar-nordic's explanation right.

**Comment by mbolivar-nordic** [comment]
Thanks! Sorry to require hand-holding, but I didn't write west's submodule support, and I don't know much about submodules. I really needed some help to get this failure on my desk.

Here is what I've figured out so far.

By replacing `west update sof` with `west -v update sof`, I see this in the output:

```
west.manifest: running 'git submodule update --init --checkout --recursive' in /home/mbolivar/zeptest/modules/audio/sof
warning: could not look up configuration 'remote.origin.url'. Assuming this repository is its own authoritative upstream.
Submodule 'rimage' (/home/mbolivar/zeptest/modules/audio/rimage.git) registered for path 'rimage'
fatal: repository '/home/mbolivar/zeptest/modules/audio/rimage.git' does not exist
fatal: clone of '/home/mbolivar/zeptest/modules/audio/rimage.git' into submodule path '/home/mbolivar/zeptest/modules/audio/sof/rimage' failed
```

So `git submodule update --init --checkout --recursive` is what's failing.

Trying to figure out why, I found that the `git help submodule` section for the `git submodule update` command says:

```
           If the submodule is not yet initialized, and you just want to use
           the setting as stored in .gitmodules, you can automatically
           initialize the submodule with the --init option.
```

Higher up in the page, the `git submodule init` help says:

```
       init [--] [<path>...]
           Initialize the submodules recorded in the index (which were added
           and committed elsewhere) by setting submodule.$name.url in
           .git/config. It uses the same setting from .gitmodules as a
           template. If the URL is relative, it will be resolved using the
           default remote. If there is no default remote, the current
           repository will be assumed to be upstream.
           [...]
           See the add subcommand for the definition of default remote.
```

The sof module's `.gitmodules` file contains:

```
[submodule "rimage"]
	path = rimage
	url = ../rimage.git
```

And the `add` subcommand help says:

```
           The default remote is the remote of the remote-tracking branch of
           the current branch. If no such remote-tracking branch exists or the
           HEAD is detached, "origin" is assumed to be the default remote.
```

Since we detach HEAD after checking out `manifest-rev`, we do indeed have a situation where git itself is assuming "origin" is the default remote due to the above chain of events.

On a hunch, I tried these commands from `zeptest/modules/audio/sof`:

```
$ git config submodule.rimage.url https://github.com/thesofproject/rimage
$ git submodule update --checkout --recursive
```

And that worked fine.

So it looks like west can't run `git submodule update --init [..]` at all; submodule update needs to be converted into a multi-step process.

I'm guessing we will have to do something like this in the general case:

1. figure out what all the submodules are
2. set their URLs manually
3. run `git submodule update --checkout --recursive`

@marc-hb and @Mierunski, you both seem to know a lot about submodules. Does the above look reasonable to you?

**Comment by marc-hb** [comment]
The analysis seems 100% correct and the suggested fix sounds good and like it should work. I haven't yet thought and tried to find other possible way(s) to approach this.

Note tentative and "unconditional" fix #546 already submitted.

> I am a bit concerned about how --recursive will play into this.

I think it wwill be fine because git, not west initializes the first level of submodules. So they will have `origin` by default.

cc: @kkasperczyk-no

**Comment by Mierunski** [comment]
Sounds like it should work, however manually setting urls of all submodules and detecting them seems prone to errors instead of letting git handle it. I'm fine with it, but I won't be able to provide patch for it.

Proposed PR passes tox tests and is currently used in our internal CI with success.

**Comment by mbolivar-nordic** [comment]
The above works with superproject URLs that are paths, too.

For example `get_submodule_url('/some/local/repository', '../rimage.git')` is `/some/local/rimage.git`.

**Comment by marc-hb** [comment]
> I think it will be fine because git, not west initializes the first level of submodules. So they will have origin by default.

So I tested --recursive. I didn't use `west` but manually reproduced the solution described by @mbolivar-nordic above. The long story is: yes I was right to be optimistic: git does creates a default remote named `origin` in the first submodule level even when the parent repo has been stripped of everything (remotes, branches).  `submodule.rimage.url` is enough. So we have a winner:
```
git config  submodule.rimage.url https://github.com/thesofproject/rimage
```
I also found the first "bug":

> 3. run git submodule update --checkout --recursive

This is of course missing `--init`.


I also noticed the "normal" submodule process (as opposed to the tentative west solution) does this by default:
```
git config submodule.$name.active true
```
This variable is missing at the first level when setting `submodule.$name.url` manually. Its lack did not prevent any `--init` or anything to work but might as well set it anyway just for consistency with git?


An important warning: don't forget the `.git/` directory of submodules is now embedded in the parent's `.git/` directory. But there is worse: there is NO simple way to "start submodules from scratch". Even `deinit --force` does not start submodules from scratch (I haven't tried to manually delete the embedded `.git/`, that sounded too crazy)

Finally, I learned something about submodules $names. I used to believe that they were purely informative but they have an important role actually: they connect the url in `.git/config` and the path in `.gitmodules` together. They can be anything and even be changed at anytime but only as long as they're the same in these two files. In other words the `.gitmodules` file is still used even after --init.

Hope this helps.

**Comment by marc-hb** [comment]
> I agree, I would like that west doesn't interfere with whatever I decide to use as origin or any other remote for that matter.

While I'm not a fan for #546 either, note it does not interfere with any `origin` already there. It only adds one when missing.

> So even setting the origin to the Zephyr fork of sof will never work.

This is because Zephyr does not need `rimage` right now. This is really not a `west` problem and is irrelevant here. The entire purpose of any relative submodule URL is to mirror **both** parent and child or none on the same server - _if_ you need both.

I used this example because @mbolivar-nordic was asking for something easy to copy/paste and because I knew about it but it's not the perfect example. To perfect this example just replace https://github.com/zephyrproject-rtos/sof with https://github.com/thesofproject/sof in the manifest and move on.

**Comment by marc-hb** [comment]
> And then describe the limitation that if you're going to use submodules in west then those submodules must use a real url and not a relative paths.

@tejlmand I'm afraid you also missed this:

> Proposed PR passes tox tests and is currently used in our internal CI with success.

**Comment by marc-hb** [comment]
> Now, the problem is that origin is not updated in the chain of submodules as far as I can tell.
> And this could mean that any new SHAs pushed to the fork could be inaccessible when fetching origin.

The funny thing is the exact same problem happens with absolute submodules: the origin is not updated either. What makes you think the old one is always valid?

**Comment by mbolivar-nordic** [comment]
I would like to unblock this discussion. I'm starting with a regression test: https://github.com/zephyrproject-rtos/west/pull/557

Please take a look if you have time, but I will start work on the fix in parallel either way, taking into account the discussion above.
